blog/_posts/2022-12-13-tut-access-to-public-ip-from-vms-containers-using-firewalld.md

```diff
@@ -0,0 +1,52 @@
+section: Blog
+date: 2022-12-13T13:00:00
+author: David Foley
+category: tutorial
+---
+
+You are running some LXC containers on a host and you use Firewalld to forward ports from the public internet to the containers. How do you enable access to the public IP of the LXD host from the LXC containers?
+
+## TL;DR
+- Ordinairly connections from containers to to services which are reached by port forwarding on Firewalld fail, because after outputting the packet to the public interface, it never returns and therefore Firewalld cannot process the port forward.
+- The soloution is destination NAT: `sudo firewall-cmd --zone=trusted --add-rich-rule='rule family="ipv4" destination address="123.123.123.123" forward-port port="80" protocol="tcp" to-port="80" to-addr="10.10.1.20"'`
+
+## The Scenario
+You have some LXC containers running on a host, the default LXD setup creates a virtual bridge to which all the containers are connected, they have their own private network say in the 10.10.1.0/24 subnet.
+
+You use Firewalld to forward ports from the public internet to the containers. In this scenario when a container does a DNS lookup to which the answer is the public IP address of the LXD host and the container then tries to connect to say port 80 on that public IP it will fail. Why? The HTTP request is received on the input chain by firewalld, no processing is required as it appears to be destined for the public internet. Firewalld outputs the packet to the public interface of the host. Since the HTTP Proxy is not bound to the public interface, it is instead reached via port forwards in Firewalld, the connection fails. This is because Firewalld handles the port forwarding and, after outputting the packet to the public interface, it never returns and therefore Firewalld cannot process the port forward.
+
+## The Solution
+Destination NAT rules in firewalld are the solution here.
+
+I understand NAT would appear to be an obvious answer. Indeed if you simply enable masquerading on the zone which contains the container virtual network this will begin to work, this has the unintended consequence of also source NATing all incoming requests to the containers. This means client IPs will no longer visible to applications running in containers.
+
+Destination NAT is applied on the input chain, before the routing decision, where it modifies the destination IP address of the packet based. In the example the diagram describes Firewalld recognises that the destination IP for the HTTP request is the public IP of the host. After this it takes the packet and changes the destination IP address to the internal IP address of the Web Proxy Container.
+
+To setup DST-NAT on the example host, you would follow this procedure:
+
+1. Execute these commands on the Firewalld host:
+```
+sudo firewall-cmd --zone=trusted --add-rich-rule='rule family="ipv4" destination address="123.123.123.123" forward-port port="80" protocol="tcp" to-port="80" to-addr="10.10.1.20"'
```

Comment:
You can replace `123.123.123.123` with `192.0.2.123`. The latter in part of the [reserved example network](https://en.wikipedia.org/wiki/Reserved_IP_addresses#IPv4), `192.0.2.0/24`.

Not required. Just a suggestion.

Reply:
Happy to change it, may I suggest "203.0.113.1"? This is also part of an reserved example network according to your link. However it has the advantage of not beginning with 192 and being confused with a private address as they often begin with 192.

Thank you.